v1/tarball/image.go

```diff
@@ -150,15 +196,74 @@ func extractFileFromTar(tarPath string, filePath string) (io.ReadCloser, error)
 	return nil, fmt.Errorf("file %s not found in tar", filePath)
 }
 
-func (i *image) ConfigFile() (*v1.ConfigFile, error) {
-	return i.config, nil
-}
-
-func (i *image) UncompressedLayer(h v1.Hash) (io.ReadCloser, error) {
+func (i *uncompressedImage) UncompressedLayer(h v1.Hash) (io.ReadCloser, error) {
 	for idx, diffID := range i.config.RootFS.DiffIDs {
 		if diffID == h {
 			return extractFileFromTar(i.path, i.imgDescriptor.Layers[idx])
 		}
 	}
 	return nil, fmt.Errorf("diff id %q not found", h)
 }
+
+func (c *compressedImage) Blob(h v1.Hash) (io.ReadCloser, error) {
+	return nil, nil
+}
+
+func (c *compressedImage) Manifest() (*v1.Manifest, error) {
+	c.manifestLock.Lock()
+	defer c.manifestLock.Unlock()
+	if c.manifest != nil {
+		return c.manifest, nil
+	}
+
+	cfg, err := c.ConfigFile()
+	if err != nil {
+		return nil, err
+	}
+	b, err := json.Marshal(cfg)
+	if err != nil {
+		return nil, err
+	}
+
+	cfgHash, cfgSize, err := v1.SHA256(v1util.NopReadCloser(bytes.NewBuffer(b)))
+	if err != nil {
+		return nil, err
+	}
+
+	c.manifest = &v1.Manifest{
+		SchemaVersion: 2,
+		MediaType:     types.DockerManifestSchema2,
+		Config: v1.Descriptor{
+			MediaType: types.DockerConfigJSON,
+			Size:      cfgSize,
+			Digest:    cfgHash,
+		},
+	}
+
+	for _, p := range c.imgDescriptor.Layers {
+		l, err := extractFileFromTar(c.path, p)
+		sha, size, err := v1.SHA256(l)
+		if err != nil {
+			return nil, err
+		}
+		c.manifest.Layers = append(c.manifest.Layers, v1.Descriptor{
+			MediaType: types.DockerLayer,
+			Size:      size,
+			Digest:    sha,
+		})
+	}
+	return nil, nil
```

Comment:
`c.manifest` LOL Kudos @jonjohnsonjr

Reply:
haha yeah I fixed this one in my other branch, which adds tar.Write and was depending on this one. Screwed it up somehow trying to detangle the branches...

Reply:
Ooh that's a good one... TODO: implement `git detangle`